k2/csrc/context.cu

```diff
@@ -128,4 +128,50 @@ void GetBlockSizesForLambda2(int32_t m, int32_t n, dim3 *block_dim,
   }
 }
 
+
+void Semaphore:: Signal(ContextPtr c) {
+  DeviceType device_type = c->GetDeviceType();
+  if (device_type_ == kUnk)
+    device_type_ = device_type;
+  else
+    K2_CHECK_EQ(device_type, device_type_)
+        << "Semaphore must always be used with the same device type.";
+  if (device_type == kCuda) {
+    cudaEvent_t event;
+    cudaError_t e = cudaEventCreateWithFlags(&event, cudaEventDisableTiming);
+    K2_CHECK_CUDA_ERROR(e) << "Error creating event";
+    // Note: this stream is subject to being overridden by With(stream..), see
+    // class With.
+    cudaStream_t stream = c->GetCudaStream();
+    e = cudaEventRecord(event, stream);
+    K2_CHECK_CUDA_ERROR(e) << "Error recording event.";
+    std::unique_lock<std::mutex> lock(events_mutex_);
```

Comment:
I would like to use `std::lock_guard` instead of `unique_lock` as we don't need `defer_lock` feature of `unique_lock`, so using `std::lock_guard` would be simple and intuitive as when it locked in construction and unlocked in destruction (when out of scope)